lib/provisioning/ProvisionDispatcher.js

```diff
@@ -0,0 +1,356 @@
+'use strict'; // eslint-disable-line
+
+const async = require('async');
+const crypto = require('crypto');
+const path = require('path');
+const zookeeper = require('node-zookeeper-client');
+
+const Logger = require('werelogs').Logger;
+
+function _genRandomHexChars(len) {
+    return crypto.randomBytes(Math.ceil(len / 2))
+        .toString('hex') // convert to hexadecimal format
+        .slice(0, len);  // return required number of characters
+}
+
+class ProvisionDispatcher {
+
+    /**
+     * Create a new provision dispatcher object
+     *
+     * @constructor
+     * @param {object} zkConfig - zookeeper config object
+     * @param {string} zkConfig.endpoint - zookeeper endpoint string
+     * where provisions are managed
+     * (e.g. "localhost:2181/shared-tasks")
+     * @param {object} logConfig - logging configuration
+     * @param {string} logConfig.logLevel - logging level
+     * @param {string} logConfig.dumpLevel - dump level
+     */
+    constructor(zkConfig, logConfig) {
+        this.zkEndpoint = zkConfig.endpoint;
+        this.client = zookeeper.createClient(zkConfig.endpoint);
+        this.client.connect();
+        this.connectWaitList = [];
+        this.client.once('connected', () => {
+            this.log.debug('connected to the ZK server');
+            this.connectWaitList.forEach(cb => cb());
+            this.connectWaitList = null;
+        });
+        this.myName = null;
+        this.myLeaderName = null;
+        this.isLeader = false;
+        this.owners = null;
+        this.provisions = null;
+
+        this.log = new Logger('Backbeat:ProvisionDispatcher',
+                              { level: logConfig.logLevel,
+                                dump: logConfig.dumpLevel });
+    }
+
+    /**
+     * Add new provisions to the list of provisions to be dispatched
+     *
+     * @param {array} provisionList - list of provision items as strings
+     * @param {function} cb - callback when done
+     * @return {undefined}
+     */
+    addProvisions(provisionList, cb) {
+        this.log.debug('add new provisions', { provisionList });
+        async.series([
+            done => this._waitConnected(done),
+            done => this._populateZkHierarchy(done),
+            done => this._addProvisionZkNodes(provisionList, done),
+        ], cb);
+    }
+
+    subscribe(cb) {
+        this._waitConnected(() => {
+            this._registerLeader(err => {
+                if (err) {
+                    return cb(err);
+                }
+                return this._onLeaderChange();
+            });
+            this._registerOwner(err => {
+                if (err) {
+                    return cb(err);
```

Comment:
This is ok, because it's more of a listener callback than a completion callback, getting called whenever changes happen on the provisioning.